Libs/Hopac/Stream.fs

```diff
@@ -523,45 +523,42 @@ module Stream =
   let consumeFun f xs = iterFun f xs |> queue
   let consume xs = iter xs |> queue
 
-  type Pipelined<'x> =
-    | Value of 'x
-    | Exn of exn
-    | Done
-
-  type [<Struct>] Struct<'x1, 'x2> (x1: 'x1, x2: 'x2) =
-    member t.Get1 = x1
-    member t.Get2 = x2
-
   let mapPipelinedJob (degree: int) (x2yJ: 'x -> #Job<'y>) (xs: Stream<'x>) =
     if degree < 1 then
       failwithf "degree must be 1 or greater, given %d" degree
     elif degree = 1 then
       mapJob x2yJ xs
     else
       delay <| fun () ->
-      let inCh = Ch ()
-      let resultMb = Mailbox ()
-      let rec loop () =
-        resultMb >>=* fun rCh ->
-          rCh >>- function
-           | Value y -> Cons (y, loop ())
-           | Exn e   -> raise e
-           | Done    -> Nil
-      inCh >>= fun (s: Struct<_, _>) ->
-          let x = s.Get1
-          let rCh = s.Get2
-          Job.tryInDelay (fun () -> x2yJ x)
-            (Value >> Ch.give rCh)
-            (Exn >> Ch.give rCh)
-      |> Job.foreverServer
-      |> Job.forN degree
-      >>=. Job.tryIn (xs
-                      |> iterJob ^ fun x ->
-                           let rCh = Ch ()
-                           inCh *<- Struct (x, rCh) >>=.
-                           resultMb *<<+ (rCh :> Job<_>))
-             (fun () -> resultMb *<<+ Job.result Done)
-             (Exn >> Job.result >> Mailbox.send resultMb)
+      let inCh, outCh = Ch(), Ch()
+      let mutable usage = 0
+      let mutable closing = false
+
+      let rec loop() =
+        Alt.choose [
+          outCh ^-> function
+            | Choice1Of2 y -> usage <- usage - 1
+                              Cons (y, loop ())
+            | Choice2Of2 e -> raise e
+          (if not closing && usage < degree then
+            inCh ^=> fun x ->
+              usage <- usage + 1
+              Job.tryInDelay 
+                (fun () -> x2yJ x)
+                (Choice1Of2 >> Ch.give outCh)
+                (Choice2Of2 >> Ch.give outCh)
+              |> Job.queue
+              >>= loop
```

Comment:
Slightly hard to see that the `Job.queue`-job is bound with `>>=` due to how the code is formatted. Would be easier to read with a temporary named assignment/variable.